### PR TITLE
feat: graceful shutdown timeout

### PR DESCRIPTION
**Pull request by Lp-Francois**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/terminus/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: https://github.com/nestjs/terminus/issues/2421

## What is the new behavior?

Add a new option `gracefulShutdownTimeoutMs` that adds a delay to the application shutdown if specified.

```md
If your application requires postponing its shutdown process, this can be done by configuring the `gracefulShutdownTimeoutMs` in the `TerminusModule options`. 
This setting can prove particularly beneficial when working with an orchestrator such as Kubernetes. 
By setting a delay slightly longer than the readiness check interval, you can achieve zero downtime 
when shutting down containers.
```

## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

### CODE REVIEW

**Review by BrunnerLivio — CHANGES_REQUESTED** [review]

**Comment by BrunnerLivio** [comment]
Thanks a lot @Lp-Francois well done!

**Comment by Lp-Francois** [comment]
Requested changes done ✅  @BrunnerLivio

File: lib/terminus-options.interface.ts [hunk 1]
```diff
@@ -2,7 +2,26 @@ import { type LoggerService, type Type } from '@nestjs/common';
...
 export type ErrorLogStyle = 'pretty' | 'json';
 
+/**
+ * The Terminus module options
+ *
+ * errorLogStyle: The style of the error logger. Either 'pretty' or 'json'. Default to 'json'.
+ * logger: The logger to use. Either default logger or your own.
+ * gracefulShutdownTimeoutMs: The timeout to wait in ms before the application shuts down. Default to 0ms.
+ */
 export interface TerminusModuleOptions {
```
Comment (Tony133) [inline lib/terminus-options.interface.ts:12]: @BrunnerLivio shouldn't we add the `@publicApi` jsdoc here? 🙂

Comment (Lp-Francois) [reply lib/terminus-options.interface.ts:13]: Done ✔️

**Review by BrunnerLivio — APPROVED** [review]

**Comment by BrunnerLivio** [comment]
Great job, about to release this!

**Comment by Lp-Francois** [comment]
Nice thanks ✌️

**Comment by BrunnerLivio** [comment]
Moved the documentation from the README.md to the NestJS docs. 

https://github.com/nestjs/docs.nestjs.com/pull/2908

File: lib/graceful-shutdown-timeout/graceful-shutdown-timeout.service.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  }
+
+  async beforeApplicationShutdown(signal: string) {
+    this.logger.log(`Received termination signal ${signal}`);
+
+    if (signal === 'SIGTERM') {
+      this.logger.log(
+        `Awaiting ${this.gracefulShutdownTimeoutMs}ms before shutdown`,
+      );
+      await sleep(this.gracefulShutdownTimeoutMs);
```
Comment (rvitaliy) [inline lib/graceful-shutdown-timeout/graceful-shutdown-timeout.service.ts:38]: @BrunnerLivio you can just:
```
import { setTimeout } from 'node:timers/promises';

await setTimeout(this.gracefulShutdownTimeoutMs);
```

see: https://nodejs.org/api/timers.html#timerspromisessettimeoutdelay-value-options

Comment (rvitaliy) [reply lib/graceful-shutdown-timeout/graceful-shutdown-timeout.service.ts:38]: @BrunnerLivio sorry, wrong mention... the author of the PR is @Lp-Francois

**Pull request merged**